### DISCUSSION TITLE
Indexを張った時のネステッドループの計算量について

### DISCUSSION THREAD

**Discussion by yuyakawai0411** [kakeru-one/web-engineering-roundtable#37] [Q&A] [answered]
# 背景
Indexを張った時のネステッドプールの計算量log(N)のlogの底がどのように定まるかきになったため

**Comment by kakeru-one** [comment]
Bツリーインデックスの計算量を考えるとき、主に検索、挿入、削除の操作に注目します。Bツリーはバランスの取れた木構造であり、これが高速なアクセス時間を可能にします。計算量は、Bツリーの高さに依存し、これは木の次数（あるいは分岐数）と格納されている要素の数によって決まります。

### Bツリーの高さ

Bツリーの高さ \( h \) は、以下の式で概算されます。

$`h \approx \log_{t} n`$

ここで、$`t`$ は最小分岐数（各内部ノードが持つ子ノードの最小数）で、$`n`$ はノードの総数です。最小分岐数は、ノードが持ち得るキーの最大数と密接に関連しています。

### 操作の計算量

- **検索**: Bツリーでの検索操作は、根から始まり、対象のキーを持つノードを見つけるまで、対応する子ノードへと進んでいきます。最悪の場合、木の高さに相当する回数のノードを訪れることになります。したがって、検索の計算量は $`O(\log n)`$です。
  
- **挿入**: 新しいキーをBツリーに挿入する場合も、まず適切な位置を見つけるために検索を行います。その後、キーを挿入することができるスペースがノードにない場合は、そのノードを分割する必要があります。この操作も木の高さに比例するため、計算量は  $`O(\log n)`$です。

- **削除**: キーをBツリーから削除する場合、同じく最初にそのキーを含むノードを検索し、それからキーを削除します。キーの削除によってノードのキー数が最小値未満になった場合、再バランスを取る必要が出てくることがあります（隣接ノードからキーを借りたり、ノードの統合が発生する可能性があります）。削除の計算量も  $`O(\log n)`$です。

これらの操作の計算量が $`O(\log n)`$であることから、Bツリーは大量のデータを効率的に扱うことが可能で、データベースやファイルシステムで広く使用されています。

**Reply by kakeru-one** [reply]
![スクリーンショット 2024-04-27 15 36 36](https://github.com/ochi-sho-private-study/mysql-sandbox/assets/118090073/86cc64c4-82a9-495e-a6f1-a9bae7dff67d)

**Comment by yuyakawai0411** [comment, marked as answer]
B-treeの各ノードが持つ分岐数がlogの底になる。
分岐数はデータ数に依存するため、決まった計算式で求めることは難しい。
詳しくは以下のChatGPTの文章を参照

**Reply by kakeru-one** [reply]
詳細
- https://github.com/ochi-sho-private-study/mysql-sandbox/discussions/37#discussioncomment-9243552

**Comment by kakeru-one** [comment]
計算式をmarkdownで書く
https://docs.github.com/ja/get-started/writing-on-github/working-with-advanced-formatting/writing-mathematical-expressions
